### ISSUE TITLE
RangeError: Maximum call stack size exceeded with template literal since 4.1.0-dev.20200923

### ISSUE THREAD

**Issue by Prior99** [microsoft/TypeScript#41651]
## TypeScript Version:

`4.1.2`, `4.2.0-dev.20201123`

Oldest affected: `4.1.0-dev.20200923`
Newest non-affected: `4.1.0-dev.20200920`

## Search Terms:

RangeError, Maximum call strack size exceeded, getTypeFacts

## Code:

```
// The unused first parameter here is necessary. Without it, the bug is not triggered.
// The default `= string` here is necessary. Without it, the bug is not triggered.
export type Id<TA, TId extends string = string> = `${TId}-${TId}`;

export class A {}

export abstract class B {
    // When specifying the second type parameter here, the bug is not triggered.
    abstract get(id: Id<A>): void;

    // Both methods are needed to trigger the bug.
    update(id: Id<A>): void {
        // The Exclamation mark (`!`) below is needed to trigger the bug.
        this.get(id!);
    }
}
```

## Expected behavior:

Compiles without error or warning.

## Actual behavior:

Compiler crashes:

```
$ /home/prior/repositories/typescript-reproduce/node_modules/.bin/tsc reproduce.ts
/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:87210
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52147:30)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
    at getTypeFacts (/home/prior/repositories/typescript-reproduce/node_modules/typescript/lib/tsc.js:52207:24)
```

## Playground Link:

[here](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201123#code/PTAEBUAsFNQVwHZwM7QCagGYEsBOyAXUABwENdSBbaA6XUGXWbZUBaAYw+mWXIE8AdKADq2ApAD2cIuIA0oCbABGcAOagWbSUQK5satXXSCAsACgQEGKDTRMpOABsiAAwC8oQvoRrXDY01Wdi4ePlwhUXEpGU0CBSVQVQ0tBB1FfUNjNDNzaAAPYklcXX5iWABJNAAecABBBXAq0ALaBDRWb2xfUE8u3wA+XtBXABIAbya0AF8AWgmp6dcAbgsLAqKS0A4nUl5QOtBx6bW8wuKiUmVvUg4iHb3WACEji1B30CsRGAQvco5sJh+N0NIlUBxJO1FGVYGQKNRaPRGNAEjZkkFtLpMkYmDk3h8rjc7qAjAQABTYNAALlAVWqdQGAEoaQA3SSU1bmfHvKxPHSQUAIqQdUDkWDsdDoRSSDIGHGKNHqXIfeDENCkWgU6m0moM5mgNmU17mFUqqxQWAAUXyD0oGuwkMF5AA1qAya4AISuRlJaBOSQAdwxErsGAIMr0croCpUSu5pokLEEpK1HsZnJVJ3M0yAA)

## Related Issues:

None

**Comment by alexbchr** [comment]
I have a very similar issue with Typescript 4.1.2.

Following code gives a similar error (see in [Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAZxgTwDYFMAmcBmUIhwDkSaWAtAMYGQB26tMCRAUCymOnAGJQCqtYPAC8iGFGC0A5mw5cA4hMwBhCKgCuIWggDKnSnFEAKWpoBG6KHAA+PfoJgBKANoBdNpJiWcAQ0oKlAAV8MAQ4AG8WODhqDS0EAC44RWAVNU1tPXRKFgBfNgB6ArgAJXQQCAA3LhgAC2Aw1EkuH1psdGqoZDrJKTgG3GaWdAAPSFgYiG14FOxRUgxMADpMYEqAHlngiFCAPgADfaA):
```ts
import styled from 'styled-components'

type FrUnit = string

type GridColumnsSpec = (number | FrUnit)[]

interface GridProps {
  columns: GridColumnsSpec
}

// Remove this line and everything is fine
export const Grid = styled.div<GridProps>``
``` 

Strangely, everything compiles fine before I put the last line... Maybe it has something to do with types that are used in Styled Components?

Here is the stacktrace it gives me in VS Code (as playground seems to just hang):

```
/home/username/my-project/node_modules/typescript/lib/tsc.js:87001
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at getResolvedBaseConstraint (/home/username/my-project/node_modules/typescript/lib/tsc.js:43729:43)
    at getBaseConstraintOfType (/home/username/my-project/node_modules/typescript/lib/tsc.js:43718:34)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:37)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
    at getTypeFacts (/home/username/my-project/node_modules/typescript/lib/tsc.js:52049:24)
```

**Comment by bengry** [comment]
I'm seeing a similar error in 4.1.3, not sure if #41693 fixes it.

We don't use `styled-components` though, and I'm still trying to pinpoint the exact code that causes the error.

**Comment by TechQuery** [comment]
Found a similar error in `4.1.3` (TypeStrong/typedoc#1440)

**Comment by JakubKoralewski** [comment]
This fixed it for me by following @bengry's directions
```diff
<+>UTF-8
===================================================================
--- package-lock.json	(revision bf32182d959d2e63264e5ffa0e48d5b4f3eb6eba)
+++ package-lock.json	(date 1610454962142)
@@ -15390,9 +15390,10 @@
       }
     },
     "typescript": {
-      "version": "4.1.3",
-      "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.1.3.tgz",
-      "integrity": "sha512-B3ZIOf1IKeH2ixgHhj6la6xdwR9QrLC5d1VKeCSY4tvkqhF2eqd9O7txNlS0PO3GrBAFIdr3L1ndNwteUbZLYg=="
+      "version": "4.2.0-dev.20201211",
+      "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.2.0-dev.20201211.tgz",
+      "integrity": "sha512-k+ENtRQ+u9ZWF7URgAAHYTPSVu8imLa7a08XmZG+swJZTeOZJMrzFfklK/Wn5765RFG6kh2nnvcQaO9AtR/DGg==",
+      "dev": true
     },
     "unicode-canonical-property-names-ecmascript": {
       "version": "1.0.4",
```

**Comment by vitu0089** [comment]
I'd like to add that when I change the module in my tsconfig file from "CommonJS" to anything else, it suddenly springs to life, giving me a lot of errors I did not previously have.
